File: data-structures-algorithms/python_leetcode/0015-threeSum.py
'''
Difficulty: MEDIUM
No.15    https://leetcode.com/problems/3sum/
'''

from typing import List

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # base case
        if len(nums) == 3:
            if sum(nums) == 0:
                return [nums]
            else:
                return []
            
        # Sort the array
        nums.sort()
        result = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            right = len(nums) - 1
            left = i + 1
            while left < right:
                tempSum = nums[i] + nums[left] + nums[right]
                if tempSum < 0:
                    left += 1
                elif tempSum > 0:
                    right -= 1
                else:
                    result.append([nums[i], nums[left], nums[right]])
                    left += 1
                    while (nums[left] == nums[left - 1] and left < right):
                        left += 1
                
        return result
    
S = Solution()
print(S.threeSum([0,1,1]))
print(S.threeSum([-1,0,1,2,-1,-4]))
print(S.threeSum([0,0,0,0]))

File: dsa-courses/python-course/js-dsa-revision/0015-threeSum.js
/*
Difficulty: MEDIUM
No.15    https://leetcode.com/problems/3sum/

Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] 
such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.

Constraints:

- 3 <= nums.length <= 3000
- -10^5 <= nums[i] <= 10^5
*/

var threeSum = function (arr) {
  // base case
  if (arr.length === 3) {
    let sum = arr[0] + arr[1] + arr[2];
    return sum === 0 ? [arr] : [];
  }

  // sort the array
  arr.sort(function (a, b) {
    return a - b;
  });
  console.log(arr);

  // define pointers
  let result = [];

  // iterate through array
  for (let i = 0; i < arr.length; i++) {
    if (i !== 0 && arr[i] === arr[i - 1]) continue;

    let left = i + 1;
    let right = arr.length - 1;
    while (left < right) {
      let sum = arr[i] + arr[left] + arr[right];
      if (sum === 0) {
        result.push([arr[i], arr[left], arr[right]]);
        left++;
        right--;

        while (left < right && arr[left] === arr[left - 1]) {
          left++;
        }

        while (left < right && arr[right] === arr[right + 1]) {
          right--;
        }
      } else if (sum < 0) {
        left++;
      } else {
        right--;
      }
    }
  }

  // return result array
  return result;
};

console.log(threeSum([-1, 0, 1, 2, -1, -4])); // [[-1,-1,2],[-1,0,1]]
console.log(threeSum([0, 1, 1])); // []
console.log(threeSum([0, 0, 0])); // [0,0,0]
